File: spec/factories/photo.rb
FactoryGirl.define do
  factory :photo do
    user

    trait :with_metadata do
      filename     'filename'
      content_type 'image/jpeg'
      width        800
      height       600
      taken_at     Time.at(331)
      latitude     39.1
      longitude    -94.1
      comment      "comment"
    end

    trait :uploaded do
      blob_bucket 'test-bucket'
      blob_key    'blob/key'
    end

    factory :real_photo do
      filename     'IMG_2598.jpg'
      content_type 'image/jpeg'
      width        3264
      height       2448
      taken_at     DateTime.parse('2013-05-14T23:55:53Z')
      latitude     39.050333333
      longitude    -94.60716667
      blob_bucket  'adamstegman-photo-albums-test'
      blob_key     'IMG_2598.jpg'

      association :user, factory: :real_user

      before(:create) { |photo| upload_photo(photo) }
    end
  end
end

File: app/assets/javascripts/models/photo.js
PhotoAlbums.Photo = DS.Model.extend({
  filename: DS.attr('string'),
  contentType: DS.attr('string'),
  width: DS.attr('number'),
  height: DS.attr('number'),
  latitude: DS.attr('number'),
  longitude: DS.attr('number'),
  takenAt: DS.attr('date'),
  comment: DS.attr('string'),
  blobBucket: DS.attr('string'),
  blobKey: DS.attr('string'),
  
  album: DS.belongsTo('album'),
  blobSession: DS.belongsTo('blobSession'),

  base64Content: function() {
    var photo = this;
    this.get('blobSession').get('s3').getObject(
      {Bucket: this.get('blobBucket'), Key: this.get('blobKey')},
      function(err, data) {
        if (!err) {
          photo.set('base64Content', data.Body.toString('base64'));
        }
      }
    );
    return undefined;
  }.property('blobBucket', 'blobKey')
});
